### ISSUE TITLE
BART: Denoising Sequence-to-Sequence Pre-training for Natural Language Generation, Translation, and Comprehension

### ISSUE THREAD

**Issue by hoshutakemoto** [TeamAidemy/ds-paper-summaries#16]
Lewis Mike, Liu Yinhan, Goyal Naman, Ghazvininejad Marjan, Mohamed Abdelrahman, Levy Omer, Stoyanov Ves, Zettlemoyer Luke. 2019. “BART: Denoising Sequence-to-Sequence Pre-training for Natural Language Generation, Translation, and Comprehension.” arXiv [cs.CL]. arXiv. https://arxiv.org/abs/1910.13461.

- 自然言語生成、翻訳、理解のための事前学習済み言語モデル BART (Bidirectional and Auto-Regressive Transformer) を提案。
- BERTで用いられた Masked Language Model (MLM) を拡張し、言語理解などにも適用できるように。
- BERTの双方向TransformerとGPTの自己回帰Transformerを組み合わせ、Sequence-to-Sequence (Seq2Seq) の形にしたもの。
- 柔軟なノイズ関数が適用可能になり、より汎用的な学習が可能。
- 少ないデータと計算で新しいタスクに対してFine-Tuningが可能。
- 多様な文章を生成できるようになった。

## Abstract
> We present BART, a denoising autoencoder for pretraining sequence-to-sequence models. BART is trained by (1) corrupting text with an arbitrary noising function, and (2) learning a model to reconstruct the original text. It uses a standard Tranformer-based neural machine translation architecture which, despite its simplicity, can be seen as generalizing BERT (due to the bidirectional encoder), GPT (with the left-to-right decoder), and many other more recent pretraining schemes. We evaluate a number of noising approaches, finding the best performance by both randomly shuffling the order of the original sentences and using a novel in-filling scheme, where spans of text are replaced with a single mask token. BART is particularly effective when fine tuned for text generation but also works well for comprehension tasks. It matches the performance of RoBERTa with comparable training resources on GLUE and SQuAD, achieves new state-of-the-art results on a range of abstractive dialogue, question answering, and summarization tasks, with gains of up to 6 ROUGE. BART also provides a 1.1 BLEU increase over a back-translation system for machine translation, with only target language pretraining. We also report ablation experiments that replicate other pretraining schemes within the BART framework, to better measure which factors most influence end-task performance.

(DeepL翻訳)

本論文では、sequence-to-sequenceモデルを事前学習するためのノイズ除去オートエンコーダBARTを紹介する。BARTは、(1)テキストを任意のノイズ関数で汚染し、(2)元のテキストを再構築するモデルを学習することにより、学習される。BARTは標準的なTranformerベースのニューラル機械翻訳アーキテクチャを使用しており、その単純さにもかかわらず、BERT（双方向エンコーダによる）、GPT（左から右へのデコーダによる）、および他の多くの最近の事前学習スキームを一般化していると見なすことができる。我々は様々なノイズ除去方式を評価し、元の文の順番をランダムに入れ替える方式と、テキストを一つのマスクトークンに置き換える新しいインフィリング方式の両方によって、最高の性能を見つけることができた。BARTは、テキスト生成のために微調整された場合に特に効果的であるが、理解タスクにも有効である。GLUEやSQuADと同等の学習資源を持つRoBERTaと同等の性能を持ち、抽象的な対話、質問応答、要約タスクにおいて、最大6ROUGEの利得を持つ新しい最先端結果を達成しました。また、BARTは機械翻訳のバックトランスレーションシステムに対して、目標言語の事前学習のみで1.1BLEUの向上を実現しています。また、BARTのフレームワークで他の事前学習スキームを再現したアブレーション実験も報告し、エンドタスクの性能に最も影響を与える要因をより適切に測定しています。

## コード
https://github.com/facebookresearch/fairseq/tree/main/examples/bart

## 解決した課題/先行研究との比較
- 2019年時点、NLPタスクにおいて、Masked Language Model (MLM) による自己教師あり学習手法は著しい成功を収めているが、特定のタスクにのみフォーカスしていた。
  - 文中の、ある単語（単語系列）の予測。
  - 文章の自動生成。

- BERT
  - 双方向Transformer。
  - **文中のランダムにマスクされた単語を予測するタスク。**
  - マスクの左右の単語を考慮できる。
  - Q&Aや文の比較は得意だが、文章生成は苦手。

- GPT
  - 自己回帰Transformer。
  - **ある単語の次の単語を予測するタスク。**
  - 自己回帰モデルの時系列を考慮した文章生成や要約が得意。
  - 一方、単語の左側の単語しか考慮できないため、質問応答や文章全体の比較に課題あり。

- より汎用的なタスクに適応できるモデル開発を目指し、本稿ではBERTとGPTを組み合わせたアーキテクチャ **BART (Bidirectional and Auto-Regressive Transformer)** を提案。
  - Sequence-to-Sequence (Seq2Seq) のモデルとして、自然言語生成、翻訳、言語理解といったタスクに適用可能に。
    - Seq2Seq: EncoderとDecoderを主要コンポーネントとして持ったアーキテクチャで、文章を入力に、文章を出力する。機械翻訳や文章要約、対話生成などの託すに利用されている。

## 技術・手法のポイント

BARTでは、次の手順で自己教師ありの事前学習を行う。
1. 学習文書に対して**任意のノイズ関数を適用**し、破損した文章データを作成する。
2. 破損した文書に対して、**双方向Transformer**と**自己回帰Transformer**を組み合わせたSeq2SeqのTransformerモデルを用い、破損文章の再構築を学習する。

破損した文章データは以下の手法を比較した。
- Token Masking
  - 入力文章中のランダムなトークンをマスクトークンへ置換（BERTを踏襲）。
- Token Deletion
  - 入力文章中のランダムなトークンを削除。
  - マスクへの置換と異なり、モデルはどの位置のトークンが削除されたかを示す必要がある。
- Text Infilling
  - 入力文章からサンプリングした複数のスパンをマスクトークンへ置換。
  - サンプリングのスパン長はポアソン分布(λ = 3)より決定。
  - サンプリングが長さ0の場合、単純なマスクトークンの挿入となる。
  - SpanBERTからインスパイアされた。
- Sentence Permutation
  - 入力文章を句点で区切って文の系列に変換し、その順序をランダムに入れ替える。
- Document Rotation
  - 入力文章中のランダムなトークンが先頭になるように、文章を回転させる。
  - このタスクによって、文章の開始の判別が学習される。

<img width="611" alt="fig1" src="https://user-images.githubusercontent.com/61570506/210940815-daa979d3-15cd-4746-9054-db254362e680.png">

加えて、以下のようなFine-Tuningを施すことでタスクに特化させた。

- Sequence Classification Tasks（文の分類）
  - Fig. 3aを参照。
  - 分類する文をエンコーダとデコーダの両方に入力。
  - デコーダの最終出力を分類器の学習に入力。
- Token Classification Tasks（単語の分類）
  - SQuADのような、トークン分類タスク。
  - 質問文の全体を結合し、エンコーダとデコーダの両方に入力。
  - デコーダ最終層の隠れ層における各トークンにあたる値を単語の表現とする。
- Sequence Generation Tasks（文の生成）
  - BARTは自己回帰デコーダを備えているため、単体で要約文生成などの文章生成タスクが可能。
  - 入力文章をエンコーダに与え、デコーダによって文章の自動生成を行う。
- Machine Translation（機械翻訳）
  - Fig. 3bを参照。
  - 多言語で学習されたエンコーダを追加することで、BARTモデルを機械翻訳の事前学習済みデコーダとして用いることができる。
  - BARTのエンコーダ部の埋め込み層を異なるエンコーダ(source encoder)に置き換える。
  - 2ステップの学習
      a. source encoder、positional embedding層、self-attentionのBARTモデルの最初のprojection行列のみを更新する。
      b. 小規模なイテレーションで全体のパラメータを更新する。

<img width="563" alt="fig3" src="https://user-images.githubusercontent.com/61570506/216522550-73972097-7de7-4603-a86b-fe4d988fc37b.png">

## 評価指標

まず、上述の事前学習手法の比較を行った。

<img width="606" alt="tab1" src="https://user-images.githubusercontent.com/61570506/216527385-156470c8-0eb8-439e-9e1c-2cdece133fc4.png">

SQuAD, MNLI, ELI5, XSum, ConvAI2, CNN/DMのタスクで比較し、いくつかのタスクでstate-of-the-artを出した。

そこから次のような知見が得られた。
- タスクによって精度は大きく左右される。
- 単語のマスキングは不可欠。
- Left-to-Rightの事前学習手法は文章生成タスクに有効。
- 双方向エンコーダはSQuADタスクに不可欠。
- 事前学習で設定した目的だけが重要なファクターではない。
- BARTはELI5以外のタスクでより高い精度を示した。

次に、RoBERTa modelと同じスケールでBARTモデルの事前学習を行った。
- その際、事前学習にはtext infillingとsentence permutationを使用。

<img width="634" alt="tab2" src="https://user-images.githubusercontent.com/61570506/216527743-7b7149b9-6123-468a-b8e8-054a8de6e1ad.png">

分類タスク（Tab. 2）では、RoBERTaとおよそ同程度の精度が示された。

<img width="609" alt="tab3" src="https://user-images.githubusercontent.com/61570506/216527797-79eee534-2ddb-48b1-9f99-af78c9dad341.png">

要約の生成タスク（Tab. 3）では、CNN/DailyMailとXSumのデータセットで学習を行った。
いずれのタスクにおいてもBARTは高いスコアを出している。

以上より、BARTを用いることで、多様で一貫性のあるテキストを生成することができた。

## 残された課題・議論
将来的に次を検討する。
- 新しい文章のノイズ適用手法による事前学習の検討。
- 特定のエンドタスクに合わせた調整。

## 重要な引用
Jacob Devlin, Ming-Wei Chang, Kenton Lee, and Kristina Toutanova. 2019. “BERT: Pre-training of deep bidirectional transformers for language understanding.“ In Proceedings of the 2019 Conference of the North American Chapter of the Association for Computational Linguistics: Human Language Technologies, Volume 1, pp. 4171–4186. Association for Computational Linguistics. https://www.aclweb.org/anthology/N19-1423.

Alec Radford, Karthik Narasimhan, Tim Salimans, and Ilya Sutskever. 2018. “Improving language understanding by generative pre-training.“

## 関連論文
類似手法
Kaitao Song, Xu Tan, Tao Qin, Jianfeng Lu, and TieYan Liu. 2019. “Mass: Masked sequence to sequence pretraining for language generation.“ In International Conference on Machine Learning, 2019.

## 参考情報
- https://data-analytics.fun/2021/03/19/understanding-bart/
- https://github.com/TeamAidemy/ds-paper-summaries/issues/4
